cranelift-codegen/src/redundant_reload_remover.rs

```diff
@@ -0,0 +1,462 @@
+        None
+    }
+
+    // Create the binding `(reg, value, offset, size)` in `env`, and throw away any previous
+    // binding associated with `reg`.
+    fn bind(&mut self, reg: RegUnit, value: Value, offset: i32, size: u32) {
+        debug_assert!(self.map.len() == self.size);
+        self.map[reg as usize] = Some(SlotInfo {
+            value,
+            offset,
+            size,
+        });
+    }
+}
+
+impl<'a> Context<'a> {
+    /// Run the remover.
+    fn cleanup_func(&mut self) {
+        let mut total_regunits = 0;
+        for ru in self.reginfo.banks {
+            total_regunits += ru.units;
+        }
+        // Perform redundant-fill removal.  We can visit the EBBs in any order, since there's no
+        // context carried across EBB boundaries at present.
+        for &ebb in self.domtree.cfg_postorder().iter().rev() {
+            self.cleanup_ebb(ebb, total_regunits);
+        }
+    }
+
+    // Returns the regunit that `v` is allocated to.  Assumes that `v` actually resides in a
+    // register.
+    fn reg_of_value(&self, v: Value) -> RegUnit {
+        match self.cur.func.locations[v] {
+            ValueLoc::Reg(ru) => ru,
+            _ => panic!("reg_of_value: value isn't in a reg"),
+        }
+    }
+
+    // Returns the stack slot that `v` is allocated to.  Assumes that `v` actually resides in a
+    // stack slot.
+    fn slot_of_value(&self, v: Value) -> &StackSlotData {
+        match self.cur.func.locations[v] {
+            ValueLoc::Stack(slot) => &self.cur.func.stack_slots[slot],
+            _ => panic!("slot_of_value: value isn't in a stack slot"),
+        }
+    }
+
+    // Invalidates any binding associated with a regunit that is written by `inst`.
+    fn invalidate_regs_written_by_inst(&mut self, inst: Inst) {
+        let dst_vals: &[Value] = self.cur.func.dfg.inst_results(inst);
+        for v in dst_vals.iter() {
+            if let ValueLoc::Reg(ru) = self.cur.func.locations[*v] {
+                self.state.env.invalidate_by_reg(ru);
+            }
+        }
+    }
+
+    // This is the main EBB processing function.  It carries an `AvailEnv` forwards along the
+    // block, using it to detect redundant fills.
+    fn cleanup_ebb(&mut self, ebb: Ebb, total_regunits: u16) {
+        self.state.init(total_regunits as usize);
+        self.cur.goto_top(ebb);
+
+        while let Some(inst) = self.cur.next_inst() {
+            debug_assert!(
+                self.state.env.invariants_ok(),
+                "cleanup_ebb: env invariants not ok"
+            );
+
+            // To avoid difficulties with the borrow checker, do this in two stages.  First,
+            // examine the instruction to see if it can deleted or modified, and park the relevant
+            // information in `replacement`.  Update `env` too.  Later, use `replacement` to
+            // actually do the transformation if necessary.
+            let mut replacement: Option<(bool, Value)> = None;
```

Comment:
very light suggestion: we could use a small enum instead of the bool, to make it easier to follow. But that's so reduced in scope that it may not be worth it.